lib/spack/spack/config.py

```diff
@@ -436,6 +436,11 @@ def push_scope(self, scope: ConfigScope) -> None:
         """Add a higher precedence scope to the Configuration."""
         tty.debug(f"[CONFIGURATION: PUSH SCOPE]: {str(scope)}", level=2)
         self.scopes[scope.name] = scope
+        if "command_line" in self.scopes:
+            # TODO when we drop 3.6
+            # self.scopes.move_to_end("command_line")
+            # hacky implementation for now
+            self.scopes["command_line"] = self.remove_scope("command_line")
```

Comment:
I left a comment in the linked issue. It seem the root cause of this is in the logic of:
```
spack.bootstrap.config.ensure_bootstrap_configuration
```
so it's possibly better to fix it there, and let `Configuration` behave like a normal stack.

Maybe we also need to adjust the logic in `activate_environment/deactivate_environment`.

Reply:
I think this is still my preference unless we have a desire to violate the rule that `command_line` is always last. I mean if we want this to be a more abstract stack implementation that makes sense to me, but this is the config stack which has a declared preference order. So I think it makes sense to enforce it here.

I'm not sure about the bootstrap context manager. It would make sense to fix it there too.

Reply:
Can't we have a more generic:
```python
def place_first(self, name: str):
    if name in self.scopes:
        self.scopes[name] = self.remove_scope(name)
```
and use that at calling site?

Reply:
Thinking about this, the implementation above would solve issues with `command_line` (which are high priority), but will not resolve the same issue with directory config scopes injected with -C

Reply:
That still leaves me unsatisfied. I think if we leave an open-ended API we are bound to get more errors. IMHO config is defined as a stack + a guaranteed order, and we try to preserve the order by our calls throughout the code. I would rather see us check and/or enforce the error in the Config class since that is part of the objects contract. 

I will try to make a more complete implementation of this that can be adapted as the `include` discussion/implementation matures. The only reason I didn't do this already was because I'm concerned about how that will take place, but I suppose whatever I add here can be refactored as part of that effort.

Reply:
> IMHO config is defined as a stack + a guaranteed order

There is no guarantee of the order in the API of `Configuration`, the order is in general determined by the factory function in:

https://github.com/spack/spack/blob/a3322459fe97d8753677561162609667dc5a7a30/lib/spack/spack/config.py#L793

and by these few lines in `spack.main`:

https://github.com/spack/spack/blob/a3322459fe97d8753677561162609667dc5a7a30/lib/spack/spack/main.py#L953-L956

In that sense `command_line` is currently treated _differently_ from any other scope.

Reply:
We guarantee it through the docs: https://github.com/spack/spack/blob/da97e9eaa7e931099fd70fb551fe0ed154a3d36a/lib/spack/docs/configuration.rst?plain=1#L51-L105

And in our API:
https://github.com/spack/spack/blob/da97e9eaa7e931099fd70fb551fe0ed154a3d36a/lib/spack/spack/cmd/common/arguments.py#L527-L532

But not in our implementation. So what I'm proposing is we should encode this so it becomes impossible for the `Configuration` to break what we say it will do, and what we are relying on it to be for our own parser implementation.  I think we're being too true to the abstraction of a "stack". IMHO that is not really what we want. We want stack like behavior but with some additional assurances i.e.
```
command_line > env > include > custom > user > site > [etc]
```

Reply:
> Thinking about this, the implementation we have right now would solve issues with `command_line` (which are high priority), but will not resolve the same issue with directory config scopes injected with `-C`:
> 
> ```
> $ spack -C <dir> ...
> ```
> 
> If that's hard to fix here, I think it's fine to defer, but at least we should know the bug exists.

@alalazo I tried to fix it but I think there needs to be some formalization of how we handle scopes.  My idea was to compute weights based off the config names that could drive reordering but I felt like it wasn't robust enough. IMO the `ConfigScope` class should have an attribute to drive the ordering.  I've configured this to fix the acute issue with a method we can expand to formally fix ordering in the future.

Reply:
> IMO the ConfigScope class should have an attribute to drive the ordering. I've configured this to fix the acute issue with a method we can expand to formally fix ordering in the future.

@psakievich I have an idea how to fix this also for `-C`, which will probably lead to deleting the new method added here. I don't think `ConfigScope` should drive the ordering, but `Configuration` - the only difference is that we can attach a "priority" to the scope. I'll try to code that, and will ping you if I get to submit a separate PR.